contrib/ts/ts.1
.\"\t$OpenBSD: ts.1,v 1.6 2022/06/30 21:40:41 jmc Exp $
.\"
.\" Copyright (c) 2022 Job Snijders <job@openbsd.org>
.\"
.\" Permission to use, copy, modify, and distribute this software for any
.\" purpose with or without fee is hereby granted, provided that the above
.\" copyright notice and this permission notice appear in all copies.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
.\" WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
.\" ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
.\" ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
.\" OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.Dd June 30, 2022
.Dt TS 1
.Os
.Sh NAME
.Nm ts
.Nd timestamp input
.Sh SYNOPSIS
.Nm
.Op Fl i | s
.Op Fl m
.Op Ar format
.Sh DESCRIPTION
The
.Nm
utility prepends a timestamp to each line of standard input and writes
it to standard output.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl i
Display time elapsed since the last timestamp.
.It Fl m
Display timestamps derived from a strictly linearly increasing clock.
Without
.Fl m ,
timestamps reflect the current date and time, including time jumps if the
system time is changed.
.It Fl s
Display time elapsed since the start of the program.
.El
.Pp
The optional
.Ar format
argument controls how the timestamp is displayed, according to the conversion
specifications described in the
.Xr strftime 3
manual page.
The default format is
.Qq %b %d %H:%M:%S ;
or
.Qq %H:%M:%S
if one of the
.Fl i
or
.Fl s
options is used.
.Pp
Some additional conversion specifications are also supported
to append microsecond resolution:
.Cm %.S ,
.Cm %.s ,
and
.Cm %.T ;
which are similar to
.Cm %S ,
.Cm %s ,
and
.Cm \&%T .
.Sh EXAMPLES
.Bd -literal -offset indent
$ (echo foo; sleep 2; echo bar) | ts
Jun 28 12:13:38 foo
Jun 28 12:13:40 bar

$ ls | ts -i %.S
00.000452 CVS
00.000595 Makefile
.Ed
.Sh SEE ALSO
.Xr strftime 3
.Sh HISTORY
A
.Nm
utility first appeared in the moreutils collection by Joey Hess, and was
rewritten from scratch for
.Ox 7.2 .
.Pp
It was imported to
.Fx
by
.An -nosplit
.An Juraj Lutter Aq Mt otis@FreeBSD.org .
.Sh AUTHORS
This
.Nm
utility was written by
.An Job Snijders Aq Mt job@openbsd.org
and
.An Claudio Jeker Aq Mt claudio@openbsd.org .


NEWS
GNU diffutils NEWS                                    -*- outline -*-

* Noteworthy changes in release 3.3 (2013-03-24) [stable]

** New features

  --new-file (-N) and --unidirectional-new-file now allow comparisons to "-".
  A standard input that's closed acts like a nonexistent file.

  A file name containing spaces, double quotes, backslashes or control
  characters is now encoded in a diff header as a double-quoted C string
  literal.  The escape sequences \\, \", \a, \b, \f, \n, \r, \t, \v and
  \ooo (a three-digit octal number between 0 and 255) are used.

** Packaging

  diffutils is now designed to build with Cygwin or MinGW rather than DJGPP.
  The ms subdirectory has been removed.


* Noteworthy changes in release 3.2 (2011-09-02) [stable]

** Changes in behavior

  --ignore-file-name-case now applies at the top level too.
  For example, "diff dir inIt" might compare "dir/Init" to "inIt".

** New features

  diff and sdiff have a new option --ignore-trailing-space (-Z).

** Packaging

  The texinfo documentation no longer specifies "front-cover" or "back-cover"
  texts, so that it may now be included in Debian's "main" section.


* Noteworthy changes in release 3.1 (2011-08-10) [stable]

** Bug fixes

  diff no longer reports spurious differences merely because two entries
  in the same directory have names that compare equal in the current
  locale, or compare equal because --ignore-file-name-case was given.

* Noteworthy changes in release 3.0 (2010-05-03) [stable]

** Bug fixes

  diff once again prints the required "\ No newline at end of file" line
  when at least one input lacks a newline-at-EOF and the final hunk plus
  context-length aligns exactly with the end of the newline-lacking file.
  [bug introduced between 2.8.7 and 2.9]

** Changes in behavior

  In context-style diffs, diff prints a portion of a preceding "function"
  line for each hunk, with --show-function-line=RE (-F) or
  --show-c-function (-p).  Now, it trims leading blanks from such lines
  before extracting a prefix.  This is useful especially when a function
  line is so far indented that the name itself would be truncated or not
  included in the limited-width substring that diff appends.

  diff once again reports a difference with the diagnostic
  "Binary files A and B differ" when at least one of the files
  appears to be binary.  From 2.8.4 through diffutils-2.9, it printed
  "Files A and B differ".


* Noteworthy changes in release 2.9 (2010-02-11) [stable]

** New features

  New diff option --suppress-blank-empty.

  Bring back support for 'diff -NUM', where NUM is a number,
  even when conforming to POSIX 1003.1-2001.  This change reverts to
  the behavior of GNU diff 2.7 and earlier.  This is a change only
  when conforming to POSIX 1003.1-2001; there is no effect when
  conforming to older POSIX versions.

  This change is in response to decisions taken in the January 2005
  Austin Group standardization meeting.  For more details, please see
  "Utility Syntax Guidelines" in the Minutes of the January 2005
  Meeting <http://www.opengroup.org/austin/docs/austin_239.html>.

  sdiff now understands '1' and '2' as synonyms for 'l' and 'r'.

** Changes in behavior

  sdiff and diff3 now invoke diff, not $(bindir)/diff

** Administrivia

  New discussion and bug-reporting address: bug-diffutils@gnu.org

  updated gnulib support


* Noteworthy changes in release 2.8.7 (2004-04-13) [stable]

  Version 2.8.7 contains no user-visible changes.

User-visible changes in version 2.8.6:

* New diff3 option --strip-trailing-cr.

* With -N and -P, inaccessible empty regular files (the kind of files
  that 'patch' creates to indicate nonexistent backups) are now
  treated as nonexistent when they are in the 'backup' file position.

* If multiple SKIP values are given to cmp, e.g., 'cmp -i 10 -i 20',
  cmp now uses the maximal value instead of the last one.

* diff now omits the ".000000000" on hosts that do not support
  fractional time stamps.

Version 2.8.5 was not publicly released.

User-visible changes in version 2.8.4:

* Diff now simply prints "Files A and B differ" instead of "Binary
  files A and B differ".  The message is output if either A or B
  appears to be a binary file, and the old wording was misleading
  because it implied that both files are binary, which is not
  necessarily the case.

User-visible changes in version 2.8.3:

* New locale: en_US.

User-visible changes in version 2.8.2:

* New diff and sdiff option:
  --tabsize=COLUMNS
* If --ignore-space-change or --ignore-all-space is also specified,
  --ignore-blank-lines now considers lines to be empty if they contain
  only white space.
* More platforms now handle multibyte characters correctly when
  excluding files by name (diff -x and -X).
* New locales: hu, pt_BR.

User-visible changes in version 2.8.1:

* Documentation fixes.

User-visible changes in version 2.8:

* cmp and diff now conform to POSIX 1003.1-2001 (IEEE Std 1003.1-2001)
  if the underlying system conforms to POSIX and if the _POSIX2_VERSION
  environment variable is set to 200112.  Conformance removes support
  for 'diff -NUM', where NUM is a number.  Use -C NUM or -U NUM instead.
* cmp now supports trailing operands SKIP1 and SKIP2, like BSD cmp.
* cmp -i or --ignore-initial now accepts SKIP1:SKIP2 option value.
* New cmp option: -n or --bytes.
* cmp's old -c or --print-chars option has been renamed;
  use -b or --print-bytes instead.
* cmp now outputs "byte" rather than "char" outside the POSIX locale.
* cmp -l's index column width now adjusts to fit larger (or smaller) files.
* cmp -l -s and cmp -s -l are not allowed.  Use cmp -s or cmp -l instead.
* diff uses ISO 8601 style time stamps for output times (e.g. "2001-11-23
  16:44:36.875702460 -0800") unless in the C or POSIX locale and the
  -c style is specified.
* diff's -I and -F options use the regexp syntax of grep, not of Emacs.
* diff now accepts multiple context arguments, and uses their maximum value.

contrib/ts/ts.c
/*	$OpenBSD: ts.c,v 1.7 2022/07/06 07:59:03 claudio Exp $	*/
/*
 * Copyright (c) 2022 Job Snijders <job@openbsd.org>
 * Copyright (c) 2022 Claudio Jeker <claudio@openbsd.org>
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

#include <sys/types.h>
#include <sys/time.h>

#include <err.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>

static const char	*format = "%b %d %H:%M:%S";
static char		*buf;
static char		*outbuf;
static size_t		 bufsize;

static void		 fmtfmt(const struct timespec *);
static size_t		 format_length(const char *);
static void __dead2 usage(void);

int
main(int argc, char *argv[])
{
	int iflag, mflag, sflag;
	int ch, prev;
	struct timespec start, now, utc_offset, ts;
	clockid_t clock = CLOCK_REALTIME;

	iflag = mflag = sflag = 0;

	while ((ch = getopt(argc, argv, "ims")) != -1) {
		switch (ch) {
		case 'i':
			iflag = 1;
			format = "%H:%M:%S";
			clock = CLOCK_MONOTONIC;
			break;
		case 'm':
			mflag = 1;
			clock = CLOCK_MONOTONIC;
			break;
		case 's':
			sflag = 1;
			format = "%H:%M:%S";
			clock = CLOCK_MONOTONIC;
			break;
		default:
			usage();
		}
	}
	argc -= optind;
	argv += optind;

	if ((iflag && sflag) || argc > 1)
		usage();

	if (argc == 1)
		format = *argv;

	bufsize = format_length(format) * 10;
	if (bufsize == 0)
		bufsize = 1;
	if ((buf = calloc(1, bufsize)) == NULL)
		err(1, NULL);
	if ((outbuf = calloc(1, bufsize)) == NULL)
		err(1, NULL);

	/* Force UTC for interval calculations. */
	if (iflag || sflag) {
		if (setenv("TZ", "UTC", 1) == -1)
			err(1, "setenv UTC");
	}

	if (clock_gettime(clock, &start) == -1)
		err(1, "clock_gettime");
	if (clock_gettime(CLOCK_REALTIME, &utc_offset) == -1)
		err(1, "clock_gettime");
	timespecsub(&utc_offset, &start, &utc_offset);

	for (prev = '\n'; (ch = getchar()) != EOF; prev = ch) {
		if (prev == '\n') {
			if (clock_gettime(clock, &now) == -1)
				err(1, "clock_gettime");
			if (iflag || sflag)
				timespecsub(&now, &start, &ts);
			else if (mflag)
				timespecadd(&now, &utc_offset, &ts);
			else
				ts = now;
			fmtfmt(&ts);
			if (iflag)
				start = now;
		}
		if (putchar(ch) == EOF)
			break;
	}
	if (ferror(stdin))
		err(1, "stdin");

	if (fclose(stdout) == EOF)
		err(1, "stdout");
	free(outbuf);
	free(buf);
	return (0);
}

static void __dead2
usage(void)
{

	fprintf(stderr, "usage: %s [-i | -s] [-m] [format]\n", getprogname());
	exit(1);
}

static void
fmtfmt(const struct timespec *ts)
{
	struct tm tm;
	char *f, us[7];
	long usec;
	size_t format_len, i;

	if (localtime_r(&ts->tv_sec, &tm) == NULL)
		err(1, "localtime");

	if (ts->tv_nsec < 0 || ts->tv_nsec >= 1000000000L)
		errx(1, "invalid nanosecond value");
	usec = ts->tv_nsec / 1000;
	us[6] = '\0';
	for (i = 6; i > 0; i--) {
		us[i - 1] = (char)('0' + usec % 10);
		usec /= 10;
	}
	format_len = format_length(format);
	for (i = 0; i <= format_len; i++)
		buf[i] = format[i];
	f = buf;

	do {
		while (*f != '\0') {
			if (*f != '%') {
				f++;
				continue;
			}
			if (f[1] == '%') {
				f += 2;
				continue;
			}
			break;
		}
		if (*f == '\0')
			break;

		f++;
		if (f[0] == '.' &&
		    (f[1] == 'S' || f[1] == 's' || f[1] == 'T')) {
			size_t len, offset;

			f[0] = f[1];
			f[1] = '.';
			f += 2;
			len = format_length(f);
			offset = (size_t)(f - buf);
			if (offset > bufsize || bufsize - offset < len + 7)
				errx(1, "expanded format string too big");
			for (i = len + 1; i > 0; i--)
				f[i + 5] = f[i - 1];
			for (i = 0; i < 6; i++)
				f[i] = us[i];
			f += 6;
		}
	} while (*f != '\0');

	if (strftime(outbuf, bufsize, buf, &tm) == 0 && buf[0] != '\0')
		errx(1, "strftime");
	if (fprintf(stdout, "%s ", outbuf) < 0)
		err(1, "stdout");
}

static size_t
format_length(const char *str)
{
	size_t len;

	for (len = 0; str[len] != '\0'; len++) {
		if (len >= SIZE_MAX / 10)
			errx(1, "format string too big");
	}
	return (len);
}
